applications/StructuralMechanicsApplication/custom_conditions/moving_load_condition.cpp

```diff
@@ -0,0 +1,570 @@
+
+        bounded_matrix<double, TDim, TDim> rotation_matrix = ZeroMatrix(TDim, TDim);
+        this->CalculateRotationMatrix(rotation_matrix, r_geom);
+
+        // rotate load to local system
+        array_1d<double, TDim> local_moving_load;
+        noalias(local_moving_load) = prod(rotation_matrix, moving_load);
+
+        VectorType normal_shape_functions_vector;
+        VectorType shear_shape_functions_vector;
+        VectorType rotational_shape_functions_vector;
+
+        // if element has rotational degrees of freedom, shape functions are exact, thus no interpolation is required
+        if (block_size > TDim){
+            CalculateExactNormalShapeFunctions(normal_shape_functions_vector, local_x_coord);
+            CalculateExactShearShapeFunctions(shear_shape_functions_vector, local_x_coord);
+            CalculateExactRotationalShapeFunctions(rotational_shape_functions_vector, local_x_coord);
+        } else {
+            GeometryType::CoordinatesArrayType local_coordinates_array = ZeroVector(3);
+            local_coordinates_array[0] = local_x_coord / r_geom.Length() * 2 - 1;
+
+            r_geom.ShapeFunctionsValues(normal_shape_functions_vector, local_coordinates_array);
+            r_geom.ShapeFunctionsValues(shear_shape_functions_vector, local_coordinates_array);
+
+        }
+
+        array_1d<double, TDim> load_first_node = ZeroVector(TDim);
+        array_1d<double, TDim> load_end_node = ZeroVector(TDim);
+
+        BoundedMatrix<double, TDim, TNumNodes> local_load_matrix = ZeroMatrix(TDim, TNumNodes);
+        BoundedMatrix<double, TDim, TNumNodes> global_load_matrix = ZeroMatrix(TDim, TNumNodes);
+
+        const Matrix global_moment_matrix = CalculateGlobalMomentMatrix(rotational_shape_functions_vector, local_moving_load);
+
+        for (IndexType i_nod = 0; i_nod < TNumNodes; ++i_nod){
+            local_load_matrix(0, i_nod) = normal_shape_functions_vector[i_nod] * local_moving_load[0];
+
+            for (IndexType ii = 1; ii < TDim; ++ii){
+                local_load_matrix(ii, i_nod) = shear_shape_functions_vector[i_nod] * local_moving_load[ii];
+            }
+        }
+
+        // rotate load back to global
+        noalias(global_load_matrix) = prod(trans(rotation_matrix), local_load_matrix);
+
+        for (IndexType ii = 0; ii < TNumNodes; ++ii){
+            const IndexType base = ii * block_size;
+
+            // add load to rhs
+            for (IndexType k = 0; k < TDim; ++k){
+                const IndexType index = base + k;
+                rRightHandSideVector[index] += global_load_matrix(k, ii);
+            }
+
+            // shape index is 0 or 1, rotation is only added to first and final node
+            for (IndexType k = 0; k < block_size - TDim; ++k){
+
+                const IndexType index = base + TDim + k;
+                rRightHandSideVector[index] = global_moment_matrix(k, ii);
+            }
+        }
+    }
+    KRATOS_CATCH( "" )
+}
+
+
+template< >
```

Comment:
You can reduce this explicit instantations using c++17 if constexpr

Reply:
ah nice, I didn't know